synchro-spec/ffg.tla

```diff
@@ -0,0 +1,206 @@
+IsFinalized(checkpoint, viewVotes, justifiedCheckpoints) ==
+    /\ checkpoint \in justifiedCheckpoints
+    /\ LET validatorsWhoCastFinalizingVote == { 
+        v \in VALIDATORS: \E finalizingVote \in viewVotes:
+            /\ finalizingVote.validator = v
+            /\ LET ffgVote == finalizingVote.ffg_vote IN
+                \* L14:
+                /\ ffgVote.source = checkpoint
+                \* L15:
+                /\ ffgVote.target[2] = checkpoint[2] + 1 }
+        IN 3 * Cardinality(validatorsWhoCastFinalizingVote) >= 2 * N
+
+AreConflictingBlocks(b1, b2) ==
+    /\ Edge(b1,b2) \notin block_graph_closure
+    /\ Edge(b2,b1) \notin block_graph_closure
+
+
+Vote(validator, ffgVote) == [
+    validator |-> validator,
+    ffg_vote |-> ffgVote
+]
+
+(*
+The most notable difference in the synchronous spec is that 
+If, at current_slot, a checkpoint with a slot number i < current_slot
+has not been justified, there will never be a vote justifying it in the future,
+because the votes are always cast s.t. the tharget checkpoint slot is current_slot
+
+This allows us to compute justified_checkpoints incrementally.
+*)
+\* @type: ($checkpoint, $checkpoint, Set(Str)) => Bool;
+CastVotes(source, target, validators) ==
+    LET ffgVote == [ source |-> source, target |-> target ] IN
+    /\ UNCHANGED <<blocks, block_graph, block_graph_closure>>
+    /\ IsValidFFGVote(ffgVote)
+    /\ validators /= {}
+    /\ ffg_votes' = ffg_votes \union { ffgVote }
+    /\ votes' = votes \union { Vote(v, ffgVote): v \in validators }
+    \* The ffg vote (b1, s1) -> (b2, s2) can justify (b,s2), for all blocks b1 <= b <= b2 in the block graph
+    \* We branch, based on whether there is a quorum of the ffg vote
+    /\ 
+        IF (3 * Cardinality({v \in VALIDATORS: Vote(v, ffgVote) \in votes'}) >= 2 * N /\ source \in justified_checkpoints)
+        THEN LET newJustifiedBlocks == { b \in blocks: 
+                /\ <<target[1], b>> \in block_graph_closure
+                /\ <<b, source[1]>> \in block_graph_closure } 
+            IN justified_checkpoints' = justified_checkpoints \union {Checkpoint(b, target[2]): b \in newJustifiedBlocks}
+        ELSE UNCHANGED justified_checkpoints
+
+SlashableNodes ==
+    LET slashable_votes == { vote1 \in votes: \E vote2 \in votes:
+        \* equivocation
+        \/ /\ vote1.validator = vote2.validator
+           /\ vote1 /= vote2
+           /\ vote1.ffg_vote.target[2] = vote2.ffg_vote.target[2]
+        \* surround voting
+        \/ /\ vote1.validator = vote2.validator
+           /\ \/ vote1.ffg_vote.source[2] < vote2.ffg_vote.source[2]
+              \/ /\ vote1.ffg_vote.source[2] = vote2.ffg_vote.source[2]
+                 /\ vote1.ffg_vote.source[1].slot < vote2.ffg_vote.source[1].slot
+           /\ vote2.ffg_vote.target[2] < vote1.ffg_vote.target[2]
+    } IN { v.validator: v \in slashable_votes }
+
+MayAdvanceSlot ==
+    \E i \in Int:
+        /\ i >= current_slot
+        /\ current_slot' = i
```

Comment:
We still have `block.slot \in BlockSlots` in `IsValidBlock` and `checkpoint_slot \in CheckpointSlots` in `IsValidCheckpoint`, and pick `srcSlot\in CheckpointSlots` below.
Should this be able to advance past those?

Reply:
Oh, good point, but my conclusion is a bit different: With an explicit `current_slot` I don't think we actually need a `MAX_..._SLOT` or `BlockSlots` (except the >= 0 part)

Reply:
I've added the condition `<= MAX_BLOCK_SLOT +2` for now, but we should revisit whether we need the bounds. @konnov for visibility.